sputnikdao2/src/proposals.rs

```diff
@@ -0,0 +1,430 @@
+            _ => PromiseOrValue::Value(()),
+        }
+    }
+
+    pub(crate) fn internal_user_info(&self) -> UserInfo {
+        let account_id = env::predecessor_account_id();
+        UserInfo {
+            amount: self.token.accounts.get(&account_id),
+            account_id,
+        }
+    }
+}
+
+#[near_bindgen]
+impl Contract {
+    /// Add proposal to this DAO.
+    #[payable]
+    pub fn add_proposal(&mut self, proposal: ProposalInput) -> u64 {
+        // 0. validate bond attached.
+        // TODO: consider bond in the token of this DAO.
+        assert!(
+            env::attached_deposit() >= self.data().config.bond.0,
+            "ERR_MIN_BOND"
+        );
+
+        // 1. validate proposal.
+        // TODO: ???
+
+        // 2. check permission of caller to add proposal.
+        assert!(
+            self.data().policy.to_policy().can_execute_action(
+                self.internal_user_info(),
+                &proposal.kind,
+                &Action::AddProposal
+            ),
+            "ERR_PERMISSION_DENIED"
+        );
+
+        // 3. actually add proposal to current list.
+        let id = self.data().last_proposal_id;
+        self.data_mut().proposals.insert(&id, &proposal.into());
+        self.data_mut().last_proposal_id += 1;
+        id
+    }
+
+    /// Act on given proposal by id, if permissions allow.
+    pub fn act_proposal(&mut self, id: u64, action: Action) {
+        let mut proposal = self.data_mut().proposals.get(&id).expect("ERR_NO_PROPOSAL");
+        // Check permissions for given action.
+        assert!(
+            self.data().policy.to_policy().can_execute_action(
+                self.internal_user_info(),
+                &proposal.kind,
+                &Action::RemoveProposal
+            ),
+            "ERR_PERMISSION_DENIED"
+        );
+        let sender_id = env::predecessor_account_id();
+        // Update proposal given action. Returns true if should be updated in storage.
+        let update = match action {
+            Action::AddProposal => env::panic(b"ERR_WRONG_ACTION"),
+            Action::RemoveProposal => {
+                self.data_mut().proposals.remove(&id);
+                false
+            }
+            Action::VoteApprove | Action::VoteReject | Action::VoteRemove => {
+                assert_eq!(
+                    proposal.status,
+                    ProposalStatus::InProgress,
+                    "ERR_PROPOSAL_NOT_IN_PROGRESS"
+                );
+                let amount = if self
+                    .data()
+                    .policy
+                    .to_policy()
+                    .is_token_weighted(&proposal.kind)
+                {
+                    self.ft_balance_of(sender_id.clone().try_into().unwrap()).0
+                } else {
+                    1
+                };
+                proposal.update_votes(sender_id, Vote::from(action), amount);
```

Comment:
I don't think `update_votes` checks that the vote didn't happen before, so you can vote multiple times.